misc-utils/kill.c

```diff
@@ -103,21 +104,84 @@ struct kill_control {
 		verbose:1;
 };
 
-static void print_signal_name(int signum)
+static void print_signal_name(int signum, bool newline)
 {
 	const char *name = signum_to_signame(signum);
+	const char *eol = newline? "\n": "";
 
 	if (name) {
-		printf("%s\n", name);
+		printf("%s%s", name, eol);
 		return;
 	}
 #ifdef SIGRTMIN
 	if (SIGRTMIN <= signum && signum <= SIGRTMAX) {
-		printf("RT%d\n", signum - SIGRTMIN);
+		printf("RT%d%s", signum - SIGRTMIN, eol);
 		return;
 	}
 #endif
-	printf("%d\n", signum);
+	printf("%d%s", signum, eol);
+}
+
+static void print_signal_mask(uint64_t sigmask, const char sep)
+{
+	for (size_t i = 0; i < sizeof(sigmask) * 8; i++) {
+		if ((((uint64_t)0x1) << i) & sigmask) {
+			const int signum = i + 1;
+			print_signal_name(signum, false);
+			putchar(sep);
+		}
+	}
+}
+
+static void print_process_signal_state(pid_t pid)
+{
+	struct path_cxt *pc = NULL;
+	FILE *fp;
+	char buf[BUFSIZ];
+
+	const struct sigfield {
+		char *key;
+		char *label;
+	} sigfields[] = {
+		{ "SigPnd:\t", "Pending (thread)"  },
+		{ "ShdPnd:\t", "Pending (process)" },
+		{ "SigBlk:\t", "Blocked"           },
+		{ "SigIgn:\t", "Ignored"           },
+		{ "SigCgt:\t", "Caught"            },
+	};
+
+	pc = ul_new_procfs_path(pid, NULL);
+	if (!pc)
+		err(EXIT_FAILURE, _("failed to initialize procfs handler"));
+	fp = ul_path_fopen(pc, "r", "status");
+	if (!fp)
+		err(EXIT_FAILURE, _("cannot open /proc/%d/status"), pid);
+
+	while (fgets(buf, sizeof(buf), fp) != NULL) {
+		for (size_t i = 0; i < ARRAY_SIZE(sigfields); i++) {
+			const char *key = sigfields[i].key;
+			size_t keylen = strlen(key);
+
+			if (strncmp(buf, key, keylen) == 0) {
+				char *val = buf + keylen;
+				uint64_t sigmask;
```

Comment:
Aren't sigmasks `int`?

Reply:
At least on x86_64, int looks too small.
```
[yamato@dev64]~/var/util-linux/util-linux% cat /tmp/foo.c
int main(void) {return sizeof(int);}
[yamato@dev64]~/var/util-linux/util-linux% gcc /tmp/foo.c
[yamato@dev64]~/var/util-linux/util-linux% ./a.out; echo $?
4
[yamato@dev64]~/var/util-linux/util-linux% ./kill -l 0xffffffffffffffff | cat -n | tail
    55	RT21
    56	RT22
    57	RT23
    58	RT24
    59	RT25
    60	RT26
    61	RT27
    62	RT28
    63	RT29
    64	RT30
```

Am I missing something?

Reply:
Ah, I looked at `sigmask(3)` which is deprecated.

pthread_sigmask(3) proposes `sigset_t`. But those are 128 *bytes* big on amd64 glibc and a structure to boot.
So a `uint64_t` is probably the best to work with.